### PR TITLE
initial commit for github connector work WIP

### PR DESCRIPTION
**Pull request by steve-barnard**
`Addresses #388 `
(https://github.com/Garden-AI/garden/issues/388)

## Overview

First draft at the GitHubConnector class. Should have relative parity with the. HFConnector class. Likely needs additional thorough UAT but seems to be replicating the existing class.

Unsure if `save_session_and_metadata.py` needs to be modified to account for the new connector. Also a minor sp. correction in a docstring.

## Discussion

Pass in the repo URL and specify a specific SHA/commit if desired (see updated class). Important to note that the example repo used here https://github.com/uw-cmg/ASR_model was the same as in the video demo from the end of 2023. This being a GH repo and not the HF hub 'repo' meant that there was an additional unzipping step needed to get the serialized models; this should be visible in the attached image showing the code used.

## Testing

Open up a garden notebook:
`garden-ai notebook start tutorial_notebook.ipynb --base-image=3.10-sklearn`
Install this dev branch of garden-ai via the following command in the notebook:
`!pip install  git+https://github.com/Garden-AI/garden.git@26a32cf8c0ac519634f3fd2c242eab5190c40d5a`

Follow the code in the image below which replicated the code from the December 2023 demo from Will.

image

I followed the video demo and was able to produce the same results via the predictions with the HFConnector class and GitHubConnector class.

Also added a similar unit test to the only other test to reference the HFConnector.

## Documentation

After additional user testing and consensus that the GitHubConnector is at parity, it may be beneficial to update some additional documentation, but no noted updates are required at this time.

----
📚 Documentation preview 📚: https://garden-ai--391.org.readthedocs.build/en/391/

### CODE REVIEW

**Review by OwenPriceSkelly — COMMENTED** [review]
this is great, thank you steve! just a couple tweaks and I can pull it over the finish line :)

File: garden_ai/model_connectors/github_conn.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        self.local_dir = local_dir or "gh_model"
+        self.enable_imports = enable_imports
+        self.metadata = ModelMetadata(
+            model_identifier=self.repo_url,
+            model_repository="GitHub",
+            model_version=self.revision,
+        )
+        try:
+            repo_url_split = repo_url.split("/")
+            readme_url = f"https://raw.githubusercontent.com/{repo_url_split[-2]}/{repo_url_split[-1]}/main/README.md"
```
[the remark this answers is not in the document]
Comment (steve-barnard) [reply garden_ai/model_connectors/github_conn.py]: @OwenPriceSkelly should be resolved, I opted to replace with branch over revision to allow the user to specify the branch as well as be sure we pull down the correct ReadMe.

File: garden_ai/model_connectors/github_conn.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            self.read_me = requests.get(readme_url).text
+        except HTTPError:
+            self.read_me = ""
+        except requests.RequestException:
+            self.read_me = ""
+
+    @trackcalls
+    def stage(self) -> str:  # TODO Review this
```
[the remark this answers is not in the document]
Comment (steve-barnard) [reply garden_ai/model_connectors/github_conn.py]: Updated, thank you for the patch doc!

File: tutorial_notebook.ipynb
[the remark this answers is not in the document]
Comment (steve-barnard) [reply tutorial_notebook.ipynb:1]: Removed ❌

**Review by OwenPriceSkelly — APPROVED** [review]
LGTM! I'll go ahead and merge, thanks again Steve :)

**Pull request merged**